pkg/k8s/controllers/reconciler.go

```diff
@@ -0,0 +1,276 @@
+		if rr.PropagateErr {
+			log.Error("The error will be propageted to the controller")
+		}
+		return
+	}
+
+	if rr.Skipped {
+		log.Info("Reconciler skipped this event")
+		return
+	}
+
+	log.Info("Reconciler has applied the CR successfully")
+}
+
+// Reconcile is invoked when controller receives resource spec.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.log.WithField("resourceName", req.NamespacedName).WithField("kind", r.Kind())
+	rr := r.doReconcile(ctx, log, req)
+
+	// defer as func to avoid 'capturing' rr - we want to provide latest values to the endReconcile
+	defer func() {
+		r.endReconcile(ctx, log, *rr)
+	}()
+
+	if rr.ReconcileErr != nil {
+		if rr.PropagateErr {
+			// Controlelr infra will retry immediately.
+			return rr.ControllerRes, rr.ReconcileErr
+		}
+
+		if rr.Skipped {
+			// no further action needed. Even if failed - honor Reconcile decision.
+			return rr.ControllerRes, nil
+		}
+
+		// if we do not report error to the controller, we need to requeue for this CR
+		if rr.ControllerRes.Requeue || rr.ControllerRes.RequeueAfter > 0 {
+			// Reconcile alredy asked for requeue, honor it
+			return rr.ControllerRes, nil
+		}
+
+		// Reconcile reported an error and did not ask for a requeue.
+		// This is OK, we will requeue with incrementing intervals.
+		rr.ControllerRes.RequeueAfter = getRequeueDuration(rr.failCount)
+	}
+
+	return rr.ControllerRes, nil
+}
+
+// Reconcile is invoked when resource spec is created or updated.
+func (r *Reconciler) doReconcile(
+	ctx context.Context,
+	log logrus.FieldLogger,
+	req ctrl.Request,
+) *ReconcileResult {
+	rr := ReconcileResult{}
+	in := r.handler.CreateResource()
+	if getErr := r.Client().Get(ctx, req.NamespacedName, in); getErr != nil {
+		log.WithError(getErr).Errorf("unable to get %s CR", r.Kind())
+		// this is likely a controller permission issue
+		rr.ControllerRes.RequeueAfter = MaxRequeueInterval()
+		rr.ReconcileErr = getErr
+		return &rr
+	}
+
+	// Status changes trigger Reconcile events, ignore those if spec did not change.
+	var hash string
+	hash, rr.ReconcileErr = in.GetSpec().Hash()
+	if rr.ReconcileErr != nil {
+		log.WithError(rr.ReconcileErr).Error("failed to calculate hash")
+		// this is very likely a permanent error, no need to retry too frequently
+		// still, if new deployment happens, we want the new deployment to retry this CR
+		rr.ControllerRes.RequeueAfter = MaxRequeueInterval()
+		return &rr
+	}
+
+	rr.Skipped = !in.GetStatus().ShouldReconcile(hash, log)
+	if rr.Skipped {
+		// accurate skip reason is logged by ShouldReconcile
+		return &rr
+	}
+
+	rr = r.handler.Reconcile(ctx, log, in)
+	if rr.Skipped {
+		// do not take any action here if impl asked to skip status+hash updates (maybe CR is meant for a diff bento)
+		// endReconcile is still called - logging done inside
+		return &rr
+	}
+
+	updateErr := r.updateStatus(ctx, log, in, &rr)
```

Comment:
What happens if the custom resource doesn't have a status sub-resource?

Reply:
we want all resources using this framework to use Status - bootstrap forces it too

Reply:
if controller does not need status, it prob does not need the framework here as well. I would prefer status to be there, even though you do not know if it is needed or not. it helps troubleshooting a lot!